File: src/com/colacelli/ircbot/plugins/autoresponse/AutoResponse.java
package com.colacelli.ircbot.plugins.autoresponse;

import com.colacelli.irclib.messages.ChannelMessage;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Properties;

public class AutoResponse {
    private static final String PROPERTIES_FILE = "autoresponse.properties";

    private static AutoResponse instance;
    private Properties properties;

    private AutoResponse() {
        properties = new Properties();
    }

    public static AutoResponse getInstance() {
        if (instance == null) {
            instance = new AutoResponse();
        }

        return instance;
    }

    public void setAutoResponse(String trigger, String response) {
        loadProperties();
        properties.setProperty(trigger.toUpperCase(), response);
        saveProperties();
    }

    public String getAutoResponse(ChannelMessage message) {
        loadProperties();

        String text = message.getText();
        String response = properties.getProperty(text.toUpperCase());

        if (response != null && !response.isEmpty()) {
            response = response.replace("$nick", message.getSender().getNick());
            response = response.replace("$channel", message.getChannel().getName());
        }

        return response;
    }

    public HashMap<String, String> getAutoResponses() {

        HashMap<String, String> autoResponses = new HashMap<>();
        properties.forEach((key, value) -> {
            String response = String.valueOf(value);
            if (!response.isEmpty()) autoResponses.put(String.valueOf(key), response);
        });

        return autoResponses;
    }

    private void loadProperties() {
        try {
            FileInputStream fileInputStream = new FileInputStream(PROPERTIES_FILE);
            properties.load(fileInputStream);
        } catch (IOException e) {
            // Properties file not found
            properties = new Properties();
            saveProperties();
        }
    }

    private void saveProperties() {
        OutputStream outputStream = null;

        try {
            outputStream = new FileOutputStream(PROPERTIES_FILE);
            properties.store(outputStream, null);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                outputStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/com/colacelli/ircbot/plugins/autoresponse/AutoResponse.kt
package com.colacelli.ircbot.plugins.autoresponse

import com.colacelli.ircbot.base.PropertiesPlugin
import com.colacelli.irclib.messages.ChannelMessage
import java.util.*
import java.util.regex.PatternSyntaxException
import kotlin.collections.HashMap

class AutoResponse : PropertiesPlugin {
    var properties = Properties()

    private object Singleton {
        val instance = AutoResponse()
    }

    companion object {
        const val PROPERTIES_FILE = "auto_response.properties"

        val instance by lazy {
            Singleton.instance
        }
    }

    fun add(trigger: String, response: String) {
        properties = loadProperties(PROPERTIES_FILE)
        properties.setProperty(trigger.toLowerCase(), response)
        saveProperties(PROPERTIES_FILE, properties)
    }

    fun del(trigger: String) {
        properties = loadProperties(PROPERTIES_FILE)
        properties.remove(trigger.toLowerCase())
        saveProperties(PROPERTIES_FILE, properties)
    }

    fun get(message: ChannelMessage) : String? {
        properties = loadProperties(PROPERTIES_FILE)

        val text = message.text
        var trigger = Regex("")
        var response = ""

        properties.forEach { key, value ->
            if (response.isBlank()) {
                trigger = try {
                    Regex(key.toString())
                } catch (e: PatternSyntaxException) {
                    // Invalid Regex saved
                    Regex.fromLiteral(key.toString())
                }

                if (text.toLowerCase().matches(trigger)) {
                    response = value.toString()
                }
            }
        }

        when (trigger.pattern.isNotBlank() && response.isNotBlank()) {
            true -> {
                response = response.replace("\$nick", message.sender!!.nick)
                response = response.replace("\$channel", message.channel.name)

                try {
                    response = text.replace(trigger, response)
                } catch (e : IndexOutOfBoundsException) {
                    // $2 not found!
                }
            }
        }

        return response
    }

    fun list() : SortedMap<String, String> {
        properties = loadProperties(PROPERTIES_FILE)
        val responses = HashMap<String, String>()

        properties.forEach { key, value ->
            if (value.toString().isNotBlank()) responses[key.toString()] = value.toString()
        }
        return responses.toSortedMap()
    }
}